plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/Measurement.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.rules.model;
+
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Represents a measurement with a name and a corresponding value.
+ *
+ * @param <T> the type of the value
+ */
+public class Measurement<T extends Number & Comparable<T>> {
+    String name;
+    T value;
+
+    /**
+     * Create a Measurement from a StreamInput
+     * @param in the StreamInput to read from
+     * @throws IOException IOException
+     */
+    @SuppressWarnings("unchecked")
+    public Measurement(StreamInput in) throws IOException {
+        this.name = in.readString();
+        this.value = (T) in.readGenericValue();
+    }
+
+    /**
+     * Write Measurement to a StreamOutput
+     * @param out the StreamOutput to write
+     * @param measurement the Measurement to write
+     * @throws IOException IOException
+     */
+    public static void writeTo(StreamOutput out, Measurement<?> measurement) throws IOException {
+        out.writeString(measurement.getName());
+        out.writeGenericValue(measurement.getValue());
+    }
+
+    /**
+     * Constructs a new Measurement with input name and value.
+     *
+     * @param name  the name of the measurement
+     * @param value the value of the measurement
+     */
+    public Measurement(String name, T value) {
+        this.name = name;
+        this.value = value;
+    }
+
+    /**
+     * Returns the name of the measurement.
+     *
+     * @return the name of the measurement
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Returns the value of the measurement.
+     *
+     * @return the value of the measurement
+     */
+    public T getValue() {
+        return value;
+    }
+
+    /**
+     * Compare two measurements on the value
+     * @param other the other measure to compare
+     * @return 0 if the value that the two measurements holds are the same
+     *         -1 if the value of the measurement is smaller than the other one
+     *         1 if the value of the measurement is greater than the other one
+     */
+    @SuppressWarnings("unchecked")
+    public int compareTo(Measurement<? extends Number> other) {
+        Number otherValue = other.getValue();
+        if (value.getClass().equals(otherValue.getClass())) {
+            return value.compareTo((T) otherValue);
+        } else {
```

Comment:
Don't really like this model to use generic and then cast it back. Why not implement the Comparator interface? Or even here, take Measurement<T> other as input. That will ensure any violations are caught at compile time instead of runtime

Reply:
Hi @jainankitk , the reason the cast is needed is because Measurements can holds different types of values, like Long / Double / Integer. We need to compare the values based on the value types.

Reply:
As discussed offline, implementing compare method within MetricType should help get rid of Measurement class:

```
public enum MetricType implements Comparator<Number> {
        JVM,CPU;
        
        public int compare(Number a, Number b) {
            switch (this) {
                case JVM: return Long.compare(a.longValue(), b.longValue());
                case CPU: return Integer.compare(a.intValue(), b.intValue());
            }
            
            return 0;
        }
    }
```